### DISCUSSION TITLE
How to use node worker_threads in Nuxt 3?

### DISCUSSION THREAD

**Discussion by mihaon** [nuxt/nuxt#28888] [Questions]
I have a dynamic server-side PDF generator in the nuxt application. But since PDF generation blocks event loop I need to put it in the separate worker thread (`node:worker_threads`).

So I create the `server/routes/offers-pdf.get.js` route and the `workers/pdf.js` worker. But when I build nuxt app, the `workers/pdf.js` is missing from the `.output` directory. So I need to manually deploy it to the production server (also with a `node_modules` directory). Imports from paths like `~/utils/someFunction.js` inside `workers/pdf.js` doesn't work too.

Is there a solution to preserve the worker in the .output directory? Maybe I should import it in the `offers-pdf.get.js` route like `import worker from ~/workers/pdf.js?worker` or something similar?

### Example

**_server/routes/offers-pdf.get.js_**
```
import fs from 'node:fs'
import { Worker } from 'node:worker_threads'

export default defineEventHandler(async event => {
    const offers = await $fetch('/api/offers')

    setResponseHeaders(event, { 'Content-Type': 'application/pdf' })

    return new Promise((resolve, reject) => {
        const worker = new Worker(process.cwd() + '/workers/pdf.js', {
            workerData: {
                offers,
            },
        })
        worker.on('message', ({ file }) => {
            const absoluteFile = process.cwd() + '/' + file
            const stream = sendStream(event, fs.createReadStream(absoluteFile))
            stream.finally(() => fs.unlinkSync(absoluteFile))
            resolve(stream)
        })
        worker.on('error', reject)
        worker.on('exit', code => {
            if (code != 0) {
                reject(new Error(`PDF worker stopped with exit code ${code}`))
            }
        })
    })
})
```

_**workers/pdf.js**_
```
import fs from 'node:fs'
import path from 'node:path'
import { workerData, parentPort } from 'node:worker_threads'
import muhammara from 'muhammara'

// This import doesn't work
// import someFunc from '~/utils/someFunction.js'

const outputFileRelativePath = 'data/temp/pdf/' + Date.now() + '-' + process.hrtime.bigint().toString() + '.pdf'
const outputFileAbsolutePath = process.cwd() + '/' + outputFileRelativePath

fs.mkdirSync(path.dirname(outputFileAbsolutePath), { recursive: true })

const pdf = new muhammara.Recipe('new', outputFileAbsolutePath)
pdf.createPage('A4', 90)
pdf.text('Offers count: ' + workerData.offers.length, 100, 100)
pdf.endPDF()

parentPort.postMessage({ file: outputFileRelativePath })
```

**_package.json_**
```
{
    "name": "",
    "private": true,
    "type": "module",
    "scripts": {
        "build": "nuxt build",
        "dev": "nuxt dev",
        "generate": "nuxt generate",
        "preview": "nuxt preview",
        "postinstall": "nuxt prepare"
    },
    "dependencies": {
        "muhammara": "~4.1"
    },
    "devDependencies": {
        "vue": "~3.4",
        "vue-router": "~4.4",
        "nuxt": "~3.12"
    }
}
```

**Comment by NikolayZobnev** [comment]
faced the same problem

**Comment by michaelKurowski** [comment]
@mihaon did you find a solution?

**Reply by mihaon** [reply]
My solution is same as in the question. But I also deploy node_modules and etc. to production. So my production directory is looks like:
```
output/ - nuxt builded files
assets.private/ - images and etc. for using in pdf.js
workers/ - pdf.js worker
node_modules/
package.json
package-lock.json
```

Yes, it's a very dirty solution but it works. Well, at least something is.

**Comment by TaQuanMinhLong** [comment]
Hi, currently I crafted a hello world example for `node:worker_threads` usage in nuxt 4

https://github.com/TaQuanMinhLong/nuxt-worker-threads-example

You can define workers in server/workers, init workers in a server plugin

**Reply by saintsGrad15** [reply]
Dude! (or very possibly, Lady!)

This totally worked. The only thing I had to change for my scenario was to change `"bundle"` to `true` in the options passed to the call to `esbuild.build()` near the end of the worker_threads module definition.

My "worker" (which is to say, "Anything at all that I want to run in a `Worker()`" imports some modules from elsewhere in my project. So, to "bundle" them is to inline the imports so, apparently, I don't have to worry about importing them at all (and thus where the hell they're located after building)!

Thank you, Ma'am! (or very possibly, Sir!)

**Reply by TaQuanMinhLong** [reply]
I'm a dude 😅 glad to help anyway

**Reply by julbd** [reply]
Oh god @saintsGrad15. I read your reply, then forgot it and spent the whole day wondering why my worker was missing its dependencies and trying many fixes…

**Reply by julbd** [reply]
Well, i guess my bug is fixed now !

**Reply by julbd** [reply]
```
      await esbuild.build({
        entryPoints: workerFiles,
        outdir: serverOutputDir,
        minify: !import.meta.dev,
        platform: 'node',
        format: 'esm',
        bundle: true,
        packages: 'bundle',
        target: nuxt.options.vite.esbuild.target,
      });
```
